lib/global_uid/base.rb

```diff
@@ -29,169 +20,104 @@ def self.create_uid_tables(id_table_name, options={})
-        self.servers = self.servers.sort_by { |s| s[:rand] }
-      end
-
-      self.servers.each do |info|
-        next if info[:cx]
-
-        if info[:new?] || ( info[:retry_at] && Time.now > info[:retry_at] )
-          info[:new?] = false
-
-          connection = new_connection(info[:name], connection_timeout)
-          info[:cx]  = connection
-          info[:retry_at] = Time.now + self.global_uid_options[:connection_retry] if connection.nil?
-        end
-      end
+      self.servers ||= init_server_info
+      self.servers.each(&:connect)
+    end
 
-      self.servers
+    def self.disconnect!
+      self.servers.each(&:disconnect!) unless servers.nil?
+      self.servers = nil
     end
 
-    def self.with_connections
+    def self.with_servers
       servers = setup_connections!
-
-      if !self.global_uid_options[:per_process_affinity]
-        servers = servers.sort_by { rand } #yes, I know it's not true random.
-      end
+      servers = servers.shuffle if !!GlobalUid.configuration.per_process_affinity?
 
       raise NoServersAvailableException if servers.empty?
 
       errors = []
-      servers.each do |s|
+      servers.each do |server|
         begin
-          yield s[:cx] if s[:cx]
+          yield server if server.active?
         rescue TimeoutException, Exception => e
-          notify e, "#{e.message}"
+          GlobalUid.configuration.notifier.notify(e)
           errors << e
-          s[:cx] = nil
-          s[:retry_at] = Time.now + 1.minute
+          server.disconnect!
+          server.update_retry_at(60)
         end
       end
 
       # in the case where all servers are gone, put everyone back in.
-      if servers.all? { |info| info[:cx].nil? }
-        servers.each do |info|
-          info[:retry_at] = Time.now - 5.minutes
-        end
+      if get_connections.empty?
+        servers.each { |server| server.update_retry_at(0) }
         raise NoServersAvailableException, "Errors hit: #{errors.map(&:to_s).join(',')}"
       end
 
-      servers.map { |s| s[:cx] }.compact
-    end
-
-    def self.notify(exception, message)
-      if self.global_uid_options[:notifier]
-        self.global_uid_options[:notifier].call(exception, message)
-      end
+      servers
     end
 
     def self.get_connections
-      with_connections {}
+      return [] if servers.nil?
+      servers.map(&:connection).compact
     end
 
     def self.get_uid_for_class(klass)
-      with_connections do |connection|
-        Timeout.timeout(self.global_uid_options[:query_timeout], TimeoutException) do
-          id = connection.insert("REPLACE INTO #{klass.global_uid_table} (stub) VALUES ('a')")
-          return id
+      with_servers do |server|
+        Timeout.timeout(GlobalUid.configuration.query_timeout, TimeoutException) do
+          return server.allocator.add(klass.global_uid_table)
         end
       end
       raise NoServersAvailableException, "All global UID servers are gone!"
```

Comment:
We should either call notify here, or remove this exception. It'll never be triggered since `with_servers` raises an exception if all servers were pulled from the pool